sdk/nodejs/provisioning/docker-provision/image.ts

```diff
@@ -0,0 +1,237 @@
+	}
+
+	/**
+	 * buildBinPath create a path to output engine session binary.
+	 *
+	 * It will store it in the cache directory with a name composed
+	 * of the base engine session as constant and the engine identifier.
+	 */
+	private buildBinPath(): string {
+		const binPath = `${ this.cacheDir }/${ this.ENGINE_SESSION_BINARY_PREFIX }-${ this.imageRef.ID }`;
+
+		switch (os.platform()) {
+			case 'win32':
+				return `${ binPath }.exe`;
+			default:
+				return binPath;
+		}
+	}
+
+	/**
+	 * pullEngineSessionBin will retrieve Dagger binary from its docker image
+	 * and copy it to the local host.
+	 * This function automatically resolves host's platform to copy the correct
+	 * binary.
+	 */
+	private pullEngineSessionBin(engineSessionBinPath: string): void {
+		// Create a temporary bin file path
+		const tmpBinPath = path.join(
+			this.cacheDir,
+			`temp-${ this.ENGINE_SESSION_BINARY_PREFIX }`
+		);
+
+		const dockerRunArgs = [
+			'docker',
+			'run',
+			'--rm',
+			'--entrypoint',
+			'/bin/cat',
+			this.imageRef.Ref,
+			`/usr/bin/${ this.ENGINE_SESSION_BINARY_PREFIX }-${ os.platform() }-${ os.arch() }`,
+		];
+
+		try {
+			const fd = fs.openSync(tmpBinPath, 'w', 0o700);
+
+			execaCommandSync(dockerRunArgs.join(' '), {
+				stdout: fd,
+				stderr: 'pipe',
+				encoding: null,
+				// Kill the process if parent exit.
+				cleanup: true,
+				// Throw on error
+				reject: false,
+				timeout: 300000,
+			});
+
+			fs.closeSync(fd);
+			fs.renameSync(tmpBinPath, engineSessionBinPath);
+		} catch (e) {
+			fs.rmSync(tmpBinPath);
+
+			throw new Error(`failed to copy engine session binary: ${ e }`);
+		}
+
+		// Remove all temporary binary files
+		// Ignore current engine session binary or other files that have not be
+		// created by this SDK.
+		try {
+			const files = fs.readdirSync(this.cacheDir);
+			files.forEach((file) => {
+				const filePath = `${ this.cacheDir }/${ file }`;
+				if (filePath === engineSessionBinPath || !file.startsWith(this.ENGINE_SESSION_BINARY_PREFIX)) {
+					return;
+				}
+
+				fs.unlinkSync(filePath);
+			});
+		} catch (e) {
+			// Log the error but do not interrupt program.
+			console.error('could not clean up temporary binary files');
+		}
+	}
+
+	/**
+	 * runEngineSession execute the engine binary and set up a GraphQL client that
+	 * target this engine.
+	 */
+	private async runEngineSession(engineSessionBinPath: string, opts: ConnectOpts): Promise<Client> {
+		const engineSessionArgs = [ engineSessionBinPath, '--remote', `docker-image://${ this.imageRef.Ref }` ];
+
+		if (opts.Workdir) {
+			engineSessionArgs.push('--workdir', opts.Workdir);
+		}
+		if (opts.Project) {
+			engineSessionArgs.push('--project', opts.Project);
+		}
+
+		const { stdout, pid } = execaCommand(engineSessionArgs.join(' '),
+			{
+				stderr: opts.OutputLog || process.stderr,
+
+				// Kill the process if parent exit.
+				cleanup: true,
+			});
+
+		// Register PID to kill it later.
+		this.enginePid = pid;
```

Comment:
In retrospect I think we should probably just save the actual object returned by `execaCommand` and then use their method for killing it: https://github.com/sindresorhus/execa#kill-a-process

It's a bit nicer in that it does sigterm followed by sigkill. It also allows us to check if the process is still alive before trying to kill it, which is good to minimize the possibility of sending the signal to the wrong process (PIDs can be re-used).

Reply:
Done !